src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```
@@ -911,120 +1137,186 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
-                pool.shutdown();
-                out.flush();
 
                 long finish = System.currentTimeMillis();
                 long timeTakenInSecs = (finish-start)/1000;
                 if(timeTakenInSecs <= 0) timeTakenInSecs = 1;
-                logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                if(logger.isInfoEnabled()) {
+                    logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                }
             } finally {
-                rw.finalise();
+                try {
+                    // Once we get here, we need to finalise starting at the concurrent wrapper, 
+                    // as there are no more non-sentinel records to be added to the queue
+                    // This eventually triggers finalisation of the underlying writer when the queue empties
+                    // This is a soft shutdown, and hence we wait below for this stage to complete in normal circumstances
+                    // Note, this blocks for writerTimeoutWaitMillis trying to legitimately add the sentinel to the end of the queue
+                    // We force the sentinel to be added in the hard shutdown phase below
+                    concurrentWrapper.finalise();
+                } finally {
+                    try {
+                        final long completionStartTime = System.currentTimeMillis();
+                        // Busy wait check for finalised to be called in the RecordWriter or something is interrupted
+                        // By this stage, there are at maximum download.internal.queue.size items remaining (default 1000)
+                        while(writerThread.isAlive() 
+                               && !writerThread.isInterrupted()
+                               && !interruptFound.get() 
```

I cannot reproduce this today. It may have been a SOLR error on one of the callables. This raises the issue of offline downloads with a failed callable within writeResultsFromIndexToStream. It should throw an error to trigger DownloadService into producing an error email.